### PR TITLE
Parse stack params

### PR DESCRIPTION
**Pull request by aprasolova**
### Problem

When I `'invoke'` smart contract and get some values from its storage, whatever type it was written with, it is returned with type `ByteArray`. I need to convert it to required type.

### Solution

A function that parses `rpc.StackParam` value to the given variable according to its type:
- UInt160
- UInt256
- int64
- []byte

### CODE REVIEW

File: pkg/rpc/stack_param.go [hunk 1]
```diff
@@ -172,3 +172,32 @@ func (p *StackParam) UnmarshalJSON(data []byte) (err error) {
...
 }
+
+func (p *StackParam) TryParse(dest interface{}) error {
+	var (
+		err  error
+		ok   bool
+		data []byte
+	)
+	switch p.Type {
+	case ByteArray:
```
Comment (im-kulikov) [inline pkg/rpc/stack_param.go]: why only for ByteArray?

File: pkg/rpc/stack_param.go [hunk 2]
```diff
@@ -172,3 +172,32 @@ func (p *StackParam) UnmarshalJSON(data []byte) (err error) {
 	}
 	return
 }
+
+func (p *StackParam) TryParse(dest interface{}) error {
```
Comment (im-kulikov) [inline pkg/rpc/stack_param.go]: `dest` can be [variadic](http://golang-club.blogspot.com/2016/09/go.html#variadic-functions)

that's would be very useful

File: pkg/rpc/stack_param.go [hunk 3]
```diff
@@ -172,3 +172,32 @@ func (p *StackParam) UnmarshalJSON(data []byte) (err error) {
...
+		ok   bool
+		data []byte
+	)
+	switch p.Type {
+	case ByteArray:
+		if data, ok = p.Value.([]byte); !ok {
+			return errors.Errorf("failed to cast %s to []byte", p.Value)
+		}
+		switch dest := dest.(type) {
+		case *util.Uint160:
```
Comment (im-kulikov) [inline pkg/rpc/stack_param.go]: add Uint256..

File: pkg/rpc/stack_param_test.go [hunk 4]
```diff
@@ -103,3 +104,58 @@ func TestStackParam_UnmarshalJSON(t *testing.T) {
 		}
 	}
 }
+
+var (
+	hash, err = util.Uint160DecodeString("0bcd2978634d961c24f5aea0802297ff128724d6")
+	byts      = hash.Bytes()
```
Comment (im-kulikov) [inline pkg/rpc/stack_param_test.go]: pls rename to `data` or `bytes`

File: pkg/rpc/stack_param_test.go [hunk 5]
```diff
@@ -103,3 +104,58 @@ func TestStackParam_UnmarshalJSON(t *testing.T) {
...
+	},
+	{
+		input:  StackParam{Type: ByteArray, Value: "0bcd2978634d961c24f5aea0802297ff128724d6"},
+		result: byts,
+	},
+}
+
+func TestStackParam_TryParse(t *testing.T) {
+	const (
+		hash = "0bcd2978634d961c24f5aea0802297ff128724d6"
```
Comment (im-kulikov) [inline pkg/rpc/stack_param_test.go]: move it from `func` to global

File: pkg/rpc/stack_param.go (hunk 2 above)
Comment (im-kulikov) [reply pkg/rpc/stack_param.go]: I think it would be better to add this functionality to `[]StackParam`

File: pkg/rpc/stack_param_test.go [hunk 6]
```diff
@@ -103,3 +104,58 @@ func TestStackParam_UnmarshalJSON(t *testing.T) {
 		}
 	}
 }
+
+var (
+	hash, err = util.Uint160DecodeString("0bcd2978634d961c24f5aea0802297ff128724d6")
+	byts      = hash.Bytes()
+)
+
+var tryParseCases = []struct {
```
Comment (im-kulikov) [inline pkg/rpc/stack_param_test.go]: unused var

File: pkg/rpc/stack_param.go (hunk 2 above)
Comment (im-kulikov) [reply pkg/rpc/stack_param.go]: `p *StackParam` rename all receivers to a single view

File: pkg/rpc/stack_param_test.go (hunk 6 above)
Comment (aprasolova) [reply pkg/rpc/stack_param_test.go]: removed

File: pkg/rpc/stack_param_test.go (hunk 5 above)
Comment (aprasolova) [reply pkg/rpc/stack_param_test.go]: done

File: pkg/rpc/stack_param_test.go (hunk 4 above)
Comment (aprasolova) [reply pkg/rpc/stack_param_test.go]: removed as unused

File: pkg/rpc/stack_param.go (hunk 3 above)
Comment (aprasolova) [reply pkg/rpc/stack_param.go]: done

File: pkg/rpc/stack_param.go [hunk 7]
```diff
@@ -172,3 +173,54 @@ func (p *StackParam) UnmarshalJSON(data []byte) (err error) {
...
+		default:
+			return errors.Errorf("cannot cast stackparam of type %s to type %s", p.Type, dest)
+		}
+	default:
+		return errors.New("cannot define stackparam type")
+	}
+	return nil
+}
+
+func reverseBytes(arr []byte) []byte {
```
Comment (im-kulikov) [inline pkg/rpc/stack_param.go]: It’s already in repo.. DRY

File: pkg/rpc/stack_param_test.go [hunk 8]
```diff
@@ -103,3 +104,93 @@ func TestStackParam_UnmarshalJSON(t *testing.T) {
...
 }
+
+const (
+	hash160 = "0bcd2978634d961c24f5aea0802297ff128724d6"
+	hash256 = "7fe610b7c8259ae949accacb091a1bc53219c51a1cb8752fbc6457674c13ec0b"
+	testString = "myteststring"
+)
+
+func TestStackParam_TryParse(t *testing.T) {
+
```
Comment (im-kulikov) [inline pkg/rpc/stack_param_test.go]: Remove empty line

File: pkg/rpc/stack_param.go (hunk 1 above)
Comment (im-kulikov) [reply pkg/rpc/stack_param.go]: String, Integer, etc..

File: pkg/rpc/stack_param.go (hunk 2 above)
Comment (aprasolova) [reply pkg/rpc/stack_param.go]: Added func that parses variadic number of params.

File: pkg/util/array.go [hunk 9]
```diff
@@ -1,14 +1,14 @@
...
-	dest := make([]byte, len(b))
-	for i, j := 0, len(b)-1; i < j; i, j = i+1, j-1 {
-		dest[i], dest[j] = b[j], b[i]
+	for i := len(arr)/2 - 1; i >= 0; i-- {
+		opp := len(arr) - 1 - i
+		arr[i], arr[opp] = arr[opp], arr[i]
 	}
-	return dest
+	return arr
 }
```
Comment (im-kulikov) [inline pkg/util/array.go]: func TestArrayReverseNastya(t *testing.T) {
	data, err := hex.DecodeString("0bcd2978634d961c24f5aea0802297ff128724d6")
	if err != nil {
		t.Fatal(err)
	}
	have := ArrayReverse(data)
	want := []byte{214, 36, 135, 18, 255, 151, 34, 128, 160, 174, 245, 36, 28, 150, 77, 99, 120, 41, 205, 11}
	if bytes.Compare(have, want) != 0 {
		t.Fatalf("expected %v got %v", want, have)
	}
	have2 := ar(data)
	if bytes.Compare(have2, want) != 0 {
		t.Fatalf("expected %v got %v", want, have)
	}

	if bytes.Compare(have2, have) != 0 {
		t.Fatalf("expected %v got %v", want, have)
	}
}

Comment (im-kulikov) [reply pkg/util/array.go]: both results are the same.. not quite clear why to change..

File: pkg/rpc/stack_param.go [hunk 10]
```diff
@@ -172,3 +173,65 @@ func (p *StackParam) UnmarshalJSON(data []byte) (err error) {
 	}
 	return
 }
+
+type StackParams []StackParam
+
+func (p *StackParams) TryParseArray(vals ...interface{}) error {
```
[the remark this answers is not in the document]
Comment (aprasolova) [reply pkg/rpc/stack_param.go]: done

File: pkg/rpc/stack_param.go [hunk 11]
```diff
@@ -172,3 +173,65 @@ func (p *StackParam) UnmarshalJSON(data []byte) (err error) {
...
+			return nil
+		default:
+			return errors.Errorf("cannot cast stackparam of type %s to type %s", p.Type, dest)
+		}
+	default:
+		return errors.New("cannot define stackparam type")
+	}
+	return nil
+}
+
```
[the remark this answers is not in the document]
Comment (aprasolova) [reply pkg/rpc/stack_param.go]: done

File: pkg/rpc/stack_param.go [hunk 12]
```diff
@@ -172,3 +173,65 @@ func (p *StackParam) UnmarshalJSON(data []byte) (err error) {
...
+		case *string:
+			*dest = string(data)
+			return nil
+		default:
+			return errors.Errorf("cannot cast stackparam of type %s to type %s", p.Type, dest)
+		}
+	default:
+		return errors.New("cannot define stackparam type")
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (aprasolova) [reply pkg/rpc/stack_param.go]: fixed

**Comment by aprasolova** [comment]
> ```go
> // StackParam respresent a stack parameter.
> type StackParam struct {
> 	Type  StackParamType `json:"type"`
> 	Value interface{}    `json:"value"`
> }
> ```
> 
> typo `respresent` -> `represent`

fixed

File: pkg/rpc/neoScanBalanceGetter.go [hunk 13]
```diff
@@ -35,6 +36,7 @@ func (s NeoScanServer) GetBalance(address string) ([]*Unspent, error) {
 		return nil, errs.Wrap(err, "Failed to decode HTTP response")
 	}
 
+	spew.Dump(balance)
```
Comment (im-kulikov) [inline pkg/rpc/neoScanBalanceGetter.go]: pls, remove debug

File: pkg/rpc/neoScanBalanceGetter.go [hunk 14]
```diff
@@ -9,6 +9,7 @@ import (
 	"github.com/CityOfZion/neo-go/pkg/core/transaction"
 	"github.com/CityOfZion/neo-go/pkg/util"
 	errs "github.com/pkg/errors"
+	"github.com/davecgh/go-spew/spew"
```
Comment (im-kulikov) [inline pkg/rpc/neoScanBalanceGetter.go]: gofmt/goimport (resort imports)

File: pkg/rpc/stack_param.go [hunk 15]
```diff
@@ -172,3 +173,64 @@ func (p *StackParam) UnmarshalJSON(data []byte) (err error) {
...
+		case *util.Uint256:
+			if *dest, err = util.Uint256DecodeBytes(data); err != nil {
+				return err
+			}
+			return nil
+		case *int64:
+			bi := &big.Int{}
+			bi.SetBytes(util.ArrayReverse(data))
+			*dest = bi.Int64()
+			return nil
```
Comment (im-kulikov) [inline pkg/rpc/stack_param.go]:
```go
func bytes2int64(b []byte) int64 {
	if l := len(b); l != 8 {
		tmp := make([]byte, 8)
		copy(tmp[8-l:], util.ArrayReverse(b))
		b = tmp
	}
	i := binary.BigEndian.Uint64(b)
	return int64(i)
}
```

is more fast..

Comment (im-kulikov) [reply pkg/rpc/stack_param.go]:
```
Benchmark_b2i/bigint-8         	20000000	        66.0 ns/op	      53 B/op	       2 allocs/op
Benchmark_b2i/bytes-8          	50000000	        28.2 ns/op	       5 B/op	       1 allocs/op
```

Comment (im-kulikov) [reply pkg/rpc/stack_param.go]: but fix ArrayReverse:

```go
// ArrayReverse return a reversed version of the given byte slice.
func ArrayReverse(b []byte) []byte {
	// Protect from big.Ints that have 1 len bytes.
	if len(b) < 2 {
		return b
	}
	dest := make([]byte, len(b))
	for i, j := 0, len(b)-1; i <= j; i, j = i+1, j-1 {
		dest[i], dest[j] = b[j], b[i]
	}
	return dest
}
```

Comment (im-kulikov) [reply pkg/rpc/stack_param.go]:
```go
// switch dest := dest.(type) {
// ...
		case *int64, *int32, *int16, *int8, *int, *uint64, *uint32, *uint16, *uint8, *uint:
			i := bytes2int64(data)
			switch dest := dest.(type) {
			case *int64:
				*dest = int64(i)
			case *int32:
				*dest = int32(i)
			case *int16:
				*dest = int16(i)
			case *int8:
				*dest = int8(i)
			case *int:
				*dest = int(i)
			case *uint64:
				*dest = i
			case *uint32:
				*dest = uint32(i)
			case *uint16:
				*dest = uint16(i)
			case *uint8:
				*dest = uint8(i)
			case *uint:
				*dest = uint(i)
			}
			return nil
// ...
```

File: pkg/rpc/neoScanBalanceGetter.go (hunk 13 above)
Comment (aprasolova) [reply pkg/rpc/neoScanBalanceGetter.go]: done

File: pkg/rpc/stack_param.go [hunk 16]
```diff
@@ -172,3 +173,89 @@ func (p *StackParam) UnmarshalJSON(data []byte) (err error) {
...
+		default:
+			return errors.Errorf("cannot cast stackparam of type %s to type %s", p.Type, dest)
+		}
+	default:
+		return errors.New("cannot define stackparam type")
+	}
+	return nil
+}
+
+func bytesToInt64(b []byte) uint64 {
```
Comment (im-kulikov) [inline pkg/rpc/stack_param.go]: rename to `bytesToUint64`

**Comment by hal0x2328** [comment]
Award aprasolova 200 points
Award im-kulikov 50 points for review

**Comment by lllwvlvwlll** [comment]
https://neoscan.io/transaction/144dc06fbc54dd84165af08ddd9c01e8a720ca1bb05471dfd1af1503160f245f

**Pull request merged**